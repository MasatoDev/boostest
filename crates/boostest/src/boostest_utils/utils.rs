use anyhow::{anyhow, Result};
use std::path::{Path, PathBuf};

use glob::glob;
use std::collections::HashMap;
use std::ffi::OsStr;
use std::fs::{self, File};
use std::io::{self, prelude::*};

pub fn read(path: &Path) -> io::Result<String> {
    let mut f = File::open(path)?;
    let mut s = String::new();
    match f.read_to_string(&mut s) {
        Ok(_) => Ok(s),
        Err(e) => Err(e),
    }
}

pub fn read_matching_files(
    patterns: Vec<String>,
    out_file_name: &String,
) -> anyhow::Result<HashMap<PathBuf, String>> {
    let mut contents: HashMap<PathBuf, String> = HashMap::new();

    for pattern in &patterns {
        let result_glob = glob(pattern)?;

        for entry in result_glob {
            let path = entry?;

            if path
                .file_name()
                .unwrap_or(OsStr::new("none_file_name"))
                .to_string_lossy()
                .contains(out_file_name.as_str())
            {
                continue;
            }

            if path.is_file() {
                contents.insert(path.clone(), fs::read_to_string(path)?);
            }
        }
    }

    Ok(contents)
}

pub fn find_boostest_json_recursive(dir: PathBuf) -> anyhow::Result<PathBuf> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Some(file_name) = path.file_name() {
                if file_name == "boostest.setting.json" {
                    return Ok(path);
                }
            }
        } else if path.is_dir() {
            if let Ok(found_path) = find_boostest_json_recursive(path) {
                return Ok(found_path);
            }
        }
    }

    Err(anyhow!("boostest.json not found"))
}

/**
 * Normalize and resolve the path.
 * parent: "hoge/a/b/c", relative: "../../d/e/f" => "hoge/a/d/e/f"
 */
pub fn normalize_and_resolve_path(parent: &Path, relative: &Path) -> Result<PathBuf, String> {
    let combined_path = parent.join(relative);
    let resolved_path = combined_path
        .components()
        .fold(PathBuf::new(), |mut acc, comp| {
            match comp {
                std::path::Component::ParentDir => {
                    acc.pop();
                }
                std::path::Component::CurDir => {}
                other => acc.push(other),
            }
            acc
        });

    if resolved_path.is_absolute() {
        Ok(resolved_path)
    } else {
        Err("The resolved path is not absolute.".to_string())
    }
}
