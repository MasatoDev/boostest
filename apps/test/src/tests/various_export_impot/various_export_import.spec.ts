import { runSnapshotTest } from "../utils";

import { ComplexChipsType } from "../../types/export_decl";
import AnoExportDefaultInterface from "../../types/export_default_interface";
import AnoExportDefaultClass from "../../types/export_default_class";
import ExportDefaultClass from "../../types/export_default_class";
import ExportDefaultClassWithDecl from "../../types/export_default_class_with_decl";
import { ExportNamedDecl } from "../../types/export_named_decl";
import ExportDefaultNamedDecl from "../../types/export_default_named_decl";
import { AnoExportNamedDeclInterface } from "../../types/export_named_decl_interface";
import { AnoExportNamedDeclClass } from "../../types/export_named_decl_class";
import Mix, { MixCompTSAlias } from "../../types/mix";
import MixInterfaceSecond, {
  MixInterfaceFirst,
} from "../../types/mix_default_normal/mix_interface";
import MixInterfaceFirst2, {
  MixInterfaceSecond2,
} from "../../types/mix_default_normal/mix_interface2";
import MixTSAliasSecond, {
  MixTSAliasFirst,
} from "../../types/mix_default_normal/mix_ts_alias";
import MixTSAliasSecond2, {
  MixTSAliasFirst2,
} from "../../types/mix_default_normal/mix_ts_alias2";
import MixClassFirst, {
  MixClassSecond,
} from "../../types/mix_default_normal/mix_class";
import MixClassFirst2, {
  MixClassSecond2,
} from "../../types/mix_default_normal/mix_class2";
import { AccessorClass } from "../../types/pattern/accessor_class";

import { boostestComplexChipsType } from "./boostest_output/boostestComplexChipsType";
import { boostestAnoExportDefaultInterface } from "./boostest_output/boostestAnoExportDefaultInterface";
import { boostestAnoExportDefaultClass } from "./boostest_output/boostestAnoExportDefaultClass";
import { boostestExportDefaultClass } from "./boostest_output/boostestExportDefaultClass";
import { boostestExportDefaultClassWithDecl } from "./boostest_output/boostestExportDefaultClassWithDecl";
import { boostestExportNamedDecl } from "./boostest_output/boostestExportNamedDecl";
import { boostestExportDefaultNamedDecl } from "./boostest_output/boostestExportDefaultNamedDecl";
import { boostestExportNamedDeclInterface } from "./boostest_output/boostestExportNamedDeclInterface";
import { boostestAnoMixInterface } from "./boostest_output/boostestAnoMixInterface";
import { boostestMixCompTSAlias } from "./boostest_output/boostestMixCompTSAlias";
import { boostestMixInterfaceFirst } from "./boostest_output/boostestMixInterfaceFirst";
import { boostestMixInterfaceSecond } from "./boostest_output/boostestMixInterfaceSecond";
import { boostestMixInterfaceFirst2 } from "./boostest_output/boostestMixInterfaceFirst2";
import { boostestMixInterfaceSecond2 } from "./boostest_output/boostestMixInterfaceSecond2";
import { boostestMixTSAliasFirst } from "./boostest_output/boostestMixTSAliasFirst";
import { boostestMixTSAliasSecond } from "./boostest_output/boostestMixTSAliasSecond";
import { boostestMixTSAliasFirst2 } from "./boostest_output/boostestMixTSAliasFirst2";
import { boostestMixTSAliasSecond2 } from "./boostest_output/boostestMixTSAliasSecond2";
import { boostestMixClassFirst } from "./boostest_output/boostestMixClassFirst";
import { boostestMixClassSecond } from "./boostest_output/boostestMixClassSecond";
import { boostestMixClassFirst2 } from "./boostest_output/boostestMixClassFirst2";
import { boostestMixClassSecond2 } from "./boostest_output/boostestMixClassSecond2";
import { boostestAccessorClass } from "./boostest_output/boostestAccessorClass";
import { boostestAnoExportNamedDeclClass } from "./boostest_output/boostestAnoExportNamedDeclClass";

describe("Various Export Methods Tests", () => {
  runSnapshotTest(
    "ComplexChipsType",
    boostestComplexChipsType<ComplexChipsType>(),
  );

  runSnapshotTest(
    "AnoExportDefaultInterface",
    boostestAnoExportDefaultInterface<AnoExportDefaultInterface>({
      name: "overridden",
    }),
  );

  runSnapshotTest(
    "AnoExportDefaultClass",
    boostestAnoExportDefaultClass<AnoExportDefaultClass>(),
  );

  runSnapshotTest(
    "ExportDefaultClass",
    boostestExportDefaultClass<ExportDefaultClass>(),
  );

  runSnapshotTest(
    "ExportDefaultClassWithDecl",
    boostestExportDefaultClassWithDecl<ExportDefaultClassWithDecl>(),
  );

  runSnapshotTest(
    "ExportNamedDecl",
    boostestExportNamedDecl<ExportNamedDecl>(),
  );

  runSnapshotTest(
    "ExportDefaultNamedDecl",
    boostestExportDefaultNamedDecl<ExportDefaultNamedDecl>(),
  );

  runSnapshotTest(
    "AnoExportNamedDeclInterface",
    boostestExportNamedDeclInterface<AnoExportNamedDeclInterface>(),
  );

  runSnapshotTest(
    "AnoExportNamedDeclClass",
    boostestAnoExportNamedDeclClass<AnoExportNamedDeclClass>(),
  );

  runSnapshotTest("Mix", boostestAnoMixInterface<Mix>({ name: "mix" }));
  runSnapshotTest(
    "MixCompTSAlias",
    boostestMixCompTSAlias<MixCompTSAlias>({ name: "mixCompTSAlias" }),
  );

  runSnapshotTest(
    "MixInterfaceFirst",
    boostestMixInterfaceFirst<MixInterfaceFirst>(),
  );
  runSnapshotTest(
    "MixInterfaceSecond",
    boostestMixInterfaceSecond<MixInterfaceSecond>(),
  );

  runSnapshotTest(
    "MixInterfaceFirst2",
    boostestMixInterfaceFirst2<MixInterfaceFirst2>(),
  );
  runSnapshotTest(
    "MixInterfaceSecond2",
    boostestMixInterfaceSecond2<MixInterfaceSecond2>(),
  );

  runSnapshotTest(
    "MixTSAliasFirst",
    boostestMixTSAliasFirst<MixTSAliasFirst>(),
  );
  runSnapshotTest(
    "MixTSAliasSecond",
    boostestMixTSAliasSecond<MixTSAliasSecond>(),
  );

  runSnapshotTest(
    "MixTSAliasFirst2",
    boostestMixTSAliasFirst2<MixTSAliasFirst2>(),
  );
  runSnapshotTest(
    "MixTSAliasSecond2",
    boostestMixTSAliasSecond2<MixTSAliasSecond2>(),
  );

  runSnapshotTest("MixClassFirst", boostestMixClassFirst<MixClassFirst>());
  runSnapshotTest("MixClassSecond", boostestMixClassSecond<MixClassSecond>());

  runSnapshotTest("MixClassFirst2", boostestMixClassFirst2<MixClassFirst2>());
  runSnapshotTest(
    "MixClassSecond2",
    boostestMixClassSecond2<MixClassSecond2>(),
  );

  runSnapshotTest("AccessorClass", boostestAccessorClass<AccessorClass>());
});
