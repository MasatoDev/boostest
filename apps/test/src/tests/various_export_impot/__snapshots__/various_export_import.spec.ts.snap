// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Various Export Methods Tests AccessorClass matches snapshot 1`] = `"{"name":"test string data","age":10,"readonlyAge":10}"`;

exports[`Various Export Methods Tests AnoExportDefaultClass matches snapshot 1`] = `"{"name":"test string data","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests AnoExportDefaultInterface matches snapshot 1`] = `"{"name":"overridden","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests AnoExportNamedDeclClass matches snapshot 1`] = `"{"name":"test string data","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests AnoExportNamedDeclInterface matches snapshot 1`] = `"{"name":"test string data","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests ComplexChipsType matches snapshot 1`] = `"{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}"`;

exports[`Various Export Methods Tests ExportDefaultClass matches snapshot 1`] = `"{"name":"test string data","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests ExportDefaultClassWithDecl matches snapshot 1`] = `"{"name":"test string data","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests ExportDefaultNamedDecl matches snapshot 1`] = `"{"name":"test string data","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests ExportNamedDecl matches snapshot 1`] = `"{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}"`;

exports[`Various Export Methods Tests Mix matches snapshot 1`] = `"{"name":"mix","chips":{"name":"test string data","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}}"`;

exports[`Various Export Methods Tests MixClassFirst matches snapshot 1`] = `"{"name":"test string data"}"`;

exports[`Various Export Methods Tests MixClassFirst2 matches snapshot 1`] = `"{"name":"test string data"}"`;

exports[`Various Export Methods Tests MixClassSecond matches snapshot 1`] = `"{"name":"test string data"}"`;

exports[`Various Export Methods Tests MixClassSecond2 matches snapshot 1`] = `"{"name":"test string data"}"`;

exports[`Various Export Methods Tests MixCompTSAlias matches snapshot 1`] = `"{"name":"mixCompTSAlias","age":10,"sex":1,"short_name":"john","favorite":{"name":"Calbee Lightly Salted","price":120},"mostFav":{"name":"Pringles Sour Cream & Onion üòÄ","price":200},"func":"() => { }","undefinedKey":"undefined","anyKey":"any","nullKey":null,"optionalKey":"test string data","unknownKey":"undefined","thisKey":{},"conditionalKey":{"name":"Kettle Brand Sea Salt & Vinegar Êó•Êú¨Ë™û„ÅåÂÖ•„Çã„Å®„Å©„ÅÜÔºüü§î","price":250},"objectKey":{},"voidKey":null,"indexedKey":"test string data","intersectionKey":{"name":"Lay's Classic","price":130,"taste":"test string data"},"arrayKey":[]}"`;

exports[`Various Export Methods Tests MixInterfaceFirst matches snapshot 1`] = `"{"name":"MixInterfaceFirst"}"`;

exports[`Various Export Methods Tests MixInterfaceFirst2 matches snapshot 1`] = `"{"name":"MixInterfaceFirst2"}"`;

exports[`Various Export Methods Tests MixInterfaceSecond matches snapshot 1`] = `"{"name":"MixInterfaceSecond"}"`;

exports[`Various Export Methods Tests MixInterfaceSecond2 matches snapshot 1`] = `"{"name":"MixInterfaceSecond2"}"`;

exports[`Various Export Methods Tests MixTSAliasFirst matches snapshot 1`] = `"{"name":"MixTSAliasFirst"}"`;

exports[`Various Export Methods Tests MixTSAliasFirst2 matches snapshot 1`] = `"{"name":"MixTSAliasFirst2"}"`;

exports[`Various Export Methods Tests MixTSAliasSecond matches snapshot 1`] = `"{"name":"MixTsAliasSecond"}"`;

exports[`Various Export Methods Tests MixTSAliasSecond2 matches snapshot 1`] = `"{"name":"MixTSAliasSecond2"}"`;
